mdp/nodes/recursive_expansion_nodes.py

```diff
@@ -0,0 +1,623 @@
+
+    :param n: The order of the function to be computed at this step.
+    :type n: int
+
+    :param cur_var: The index of the current variable to be considered in the
+        recursion. This value will have to be lower than x.shape[1].
+    :type cur_var: int
+
+    :param pos: The index of the element to be computed, along the second
+        dimension of result.
+    :type pos: int
+
+    :returns: The vectorized result (along the observations) of the
+        n-th recursion step of the cur_var-th variable.
+    :rtype: numpy.ndarray
+
+    .. admonition:: Reference
+
+        https://en.wikipedia.org/wiki/Chebyshev_polynomials
+    """
+    return 2. * x[:, cur_var] * result[:, pos-1] - result[:, pos-2]
+
+
+# collect the recusions and set the corresponding domains
+recfs = {'standard_poly': (init_standard_poly, recf_standard_poly,
+                           -float('Inf'), float('Inf')),
+         'legendre_poly': (init_legendre_poly, recf_legendre_poly,
+                           -1, 1),
+         'legendre_rational': (init_legendre_rational,
+                               recf_legendre_rational, 0, float('Inf')),
+         'chebyshev_poly': (init_chebyshev_poly,
+                            recf_chebyshev_poly, -1, 1)}
+
+
+def process(off, leadVar, leadDeg, deg, pos, result, todoList):
+    """Computes an inplace not exactly tensorproduct of the given
+    basis functions (only the simplex of the single variable bases,
+    as we limit the multi variable polynomials by a given maximum degree).
+    """
+    for v in range(leadVar):
+        l = deg-leadDeg
+        k = 1+deg*v
+        for d in range(1, deg-leadDeg+1):
+            l = deg+1-leadDeg-d
+            result[:, pos:pos+l] = result[:, 1+off:1+off+l] * result[:, k:k+1]
+            if l > 1 and leadVar > 0:
+                todoList.append((pos-1, v, leadDeg+d))
+
+            pos += l
+            k += 1
+            l -= 1
+
+    return pos
+
+
+class RecursiveExpansionNode(PolynomialExpansionNode):
+    """Recursively computable (orthogonal) expansions.
+
+    .. attribute:: lower
+
+        The lower bound of the domain, on which the recursion
+        function sequence selected, is defined or orthogonal.
+
+    .. attribute:: upper
+
+        The upper bound of the domain, on which the recursion
+        function sequence selected, is defined or orthogonal.
+    """
+
+    def __init__(self, degree=1, recf='standard_poly', check=False,
+                 with0=True, input_dim=None, dtype=None):
+        """Initialize a RecursiveExpansionNode.
+
+        :param degree: The maximum order of the recursive expansion.
+            The dimension of the return for single variable inputs will be
+            equal to this value if with0 == False.
+        :type degree: int
+
+        :param recf: Must be in ['standard_poly', 'legendre_poly',
+            'legendre_rational', 'chebyshev_poly'] or a tuply similar
+            to those in the recfs dictionary in this module.
```

Comment:
Drop a note here how the user would use his own recursion formula.

Reply:
Done